### ISSUE TITLE
add alloc_array function

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#440]
Stumpless includes wrappers for memory allocation to allow a customized memory allocator to be used. The primary function used for this is the `alloc_mem` function, which calls the underlying allocator, which is `malloc` by default.

A frequent idiom in the code base is to use `alloc_mem` with the result of a multiplication of a count of items by their size. Consider this line taken from `src/entry.c` in the `stumpless_copy_entry` function:

```c
copy->elements = alloc_mem( entry->element_count * sizeof( element_copy ) );
```

There are two things here that can be improved. First and most importantly, there is not overflow protection on the multiplication. If `element_count` causes overflow when multiplied with the size of the element copy, this memory block could be too small, and will almost certainly result in heap corruption. Second, this pattern is used in other places in the code base, and adding overflow checks to each one leaves the potential to miss one. It would be better to consolidate this logic.

## General Approach
There are a few details left out of the following approach, for you to fill in as you encounter them. If you find you need help, please ask here or on the [project gitter](https://gitter.im/stumpless/community) and someone can help you get past the stumbling block.

First, read [this](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions) section of the development documentation on adding new functions. Once you understand how this process works, start defining the new function.

Add a new function to `src/memory.c` named `alloc_array` with the following signature:

```c
void *
alloc_array( size_t item_count, size_t item_size );
```

This function will perform a multiplication overflow check, and if this passes call `alloc_mem` with the correct size. If it fails, it will raise a memory allocation failure just as `alloc_mem` does and return NULL.

If you need an example of overflow detection, have a look at [musl's implementation of `calloc`](https://git.musl-libc.org/cgit/musl/tree/src/malloc/calloc.c), which includes the same check. While `alloc_array` does not need to zero memory, the overflow detection will be the same.

Once you've made this change, swap out the usage in `src/entry.c` mentioned above, and make sure that all tests continue to pass.

Finally, add a new check to ensure that the overflow detection actually works. This should go in the test suite for the function that you updated to use the new function, which is in `test/function/entry.cpp`. The simplest way to achieve this is to create an entry, and manually update the `element_count` field to be something sufficiently large, like `SIZE_MAX`.

**Comment by aadit-n3rdy** [comment]
Hi, can I be assigned this issue?

**Comment by goatshriek** [comment]
Closed by #458.
